docs/typescript.md

```diff
@@ -0,0 +1,186 @@
+---
+id: typescript
+title: Using TypeScript with React Native
+---
+
+[TypeScript][ts] is a language which extends JavaScript to add type definitions, a lot like [Flow][flow] which React Native is built in. React Native supports TypeScript and Flow out of the box by default.
+
+## Getting Started with TypeScript
+
+If you're starting a new project, you can use the [TypeScript template][ts-template] to get you there faster:
+
+```sh
+react-native init MyAwesomeTSProject --template typescript
+```
+
+Or if you're using expo:
+
+```sh
+npm install -g expo-cli
+expo init MyAwesomeTSProject
+```
+
+and pick one of the TypeScript templates.
+
+## Adding TypeScript to an Existing Project
+
+1. Add TypeScript, and the React Native and Jest types to your project. When you were using JavaScript these could have been added for you by your editor.
+
+```sh
+yarn add typescript @types/jest @types/react @types/react-native @types/react-test-renderer
+# or for npm
+npm install --save-dev @types/jest @types/react @types/react-native @types/react-test-renderer
+```
+
+2. Add a TypeScript config file. Create a `tsconfig.json` in the root of your project:
+
+```json
+{
+  "compilerOptions": {
+    "allowJs": true,
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true,
+    "isolatedModules": true,
+    "jsx": "react",
+    "lib": ["es6"],
+    "moduleResolution": "node",
+    "noEmit": true,
+    "strict": true,
+    "target": "esnext"
+  },
+  "exclude": [
+    "node_modules",
+    "babel.config.js",
+    "metro.config.js",
+    "jest.config.js"
+  ]
+}
+```
+
+3. Create a `jest.config.js` file to configure Jest to use TypeScript
+
+```js
+module.exports = {
+  preset: 'react-native',
+  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
```

Comment:
Jest has it by default, however the order is different so this may be a bit more performant in resolving ts files

Reply:
I thought this too, but the example test doesn't pass when you have it removed

```
orta@Perl ~/d/r/t/js> mv jest.config.js _something
orta@Perl ~/d/r/t/js> yarn jest
yarn run v1.17.3
$ /Users/orta/dev/react-native/tmp/js/node_modules/.bin/jest
 FAIL  __tests__/App-test.js
  ✕ renders correctly (15ms)

  ● renders correctly

    Invariant Violation: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.

      at invariant (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:55:15)
      at createFiberFromTypeAndProps (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:2056:11)
      at createFiberFromElement (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:2077:15)
      at reconcileSingleElement (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:4629:23)
      at reconcileChildFibers (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:4686:35)
      at reconcileChildren (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:6508:28)
      at updateHostRoot (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:6920:5)
      at beginWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:7745:14)
      at performUnitOfWork (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11413:12)
      at workLoop (node_modules/react-test-renderer/cjs/react-test-renderer.development.js:11445:24)

  console.error node_modules/react/cjs/react.development.js:188
    Warning: React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: object.
```

Reply:
That should most definitely be fixed...